File: prob2.ml
open List
open Printf

let max_value : int = 4000000;;
let tally : int ref = ref 0;;

let fibo_sum : unit =
  let rec fibo_sequence (input : int list) : int list =
    match input with
      | hd1::hd2::tl ->
          let next : int = hd1 + hd2 in
          if next <= max_value
          then (fibo_sequence (next::input))
          else input
      | _ -> []
  in
  let fibo_list = fibo_sequence [2;1] in
  let fibo_check (elt : int) : unit =
    if elt mod 2 == 0
    then tally := !tally + elt
    else ()
  in
  List.iter fibo_check fibo_list
;;

Printf.printf "Sum of even elts in fibonacci sequence of max %d is: %d\n" max_value !tally;;

File: prob2.cpp
#include <iostream>
#include <cmath>
#include <cstdlib>
#include <vector>

#define MAX_VALUE 4000000

class fibo_vector
{
    protected:
        std::vector<unsigned int> numbers;
        
    public:
        unsigned int tally;

        fibo_vector(unsigned int max)
        {
            numbers.push_back(1);
            numbers.push_back(2);

            unsigned int current = 0;

            while (current <= max)
            {
                if (current != 0)
                    numbers.push_back(current);
                    
                unsigned int last = (numbers.size() - 1);
                if (numbers[last] % 2 == 0)
                    this->tally += numbers[last];
                    
                current = numbers[last] + numbers[last - 1];
            }
            
        }
};
        
int main(int argc, char *argv[])
{    
    fibo_vector* vector = new fibo_vector(MAX_VALUE);

    std::cout << "Sum of even elts in fibonacci sequence of max "
    << MAX_VALUE << " is: " << vector->tally << std::endl;
    
    return 0;
}

